flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNHighAvailabilityITCase.java

```diff
@@ -18,198 +18,205 @@
+
+		final JobGraph job = createJobGraph();
+
+		final JobID jobId = submitJob(restClusterClient, job);
+
+		final ApplicationId id = restClusterClient.getClusterId();
+
+		final long retryTimeout = 100L;
+		waitUntilJobIsRunning(restClusterClient, jobId, retryTimeout);
+
+		killApplicationMaster(yarnClusterDescriptor.getYarnSessionClusterEntrypoint());
+
+		final YarnClient yarnClient = getYarnClient();
+		Assert.assertNotNull(yarnClient);
+
+		while (yarnClient.getApplicationReport(id).getCurrentApplicationAttemptId().getAttemptId() < 2) {
+			Thread.sleep(retryTimeout);
+		}
+
+		waitUntilJobIsRunning(restClusterClient, jobId, retryTimeout);
+
+		yarnClient.killApplication(id);
+
+		while (yarnClient.getApplications(EnumSet.of(YarnApplicationState.KILLED)).size() == 0 &&
+			yarnClient.getApplications(EnumSet.of(YarnApplicationState.FINISHED)).size() == 0) {
+			Thread.sleep(retryTimeout);
+		}
+	}
+
+	@Nonnull
+	private YarnClusterDescriptor setupYarnClusterDescriptor() {
+		final Configuration flinkConfiguration = new Configuration();
+		flinkConfiguration.setString(YarnConfigOptions.APPLICATION_ATTEMPTS, "10");
+		flinkConfiguration.setString(HighAvailabilityOptions.HA_MODE, "zookeeper");
+		flinkConfiguration.setString(HighAvailabilityOptions.HA_STORAGE_PATH, storageDir);
+		flinkConfiguration.setString(HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, zkServer.getConnectString());
+		flinkConfiguration.setInteger(HighAvailabilityOptions.ZOOKEEPER_SESSION_TIMEOUT, 1000);
+
+		final int minMemory = 100;
+		flinkConfiguration.setInteger(ResourceManagerOptions.CONTAINERIZED_HEAP_CUTOFF_MIN, minMemory);
+
+		return createYarnClusterDescriptor(flinkConfiguration);
+	}
+
+	private RestClusterClient<ApplicationId> deploySessionCluster(YarnClusterDescriptor yarnClusterDescriptor) throws ClusterDeploymentException {
+		final int containerMemory = 256;
+		final ClusterClient<ApplicationId> yarnClusterClient = yarnClusterDescriptor.deploySessionCluster(
+			new ClusterSpecification.ClusterSpecificationBuilder()
+				.setMasterMemoryMB(containerMemory)
+				.setTaskManagerMemoryMB(containerMemory)
+				.setSlotsPerTaskManager(1)
+				.createClusterSpecification());
+
+		assertThat(yarnClusterClient, is(instanceOf(RestClusterClient.class)));
+		return (RestClusterClient<ApplicationId>) yarnClusterClient;
+	}
+
+	private JobID submitJob(RestClusterClient<ApplicationId> restClusterClient, JobGraph job) throws InterruptedException, java.util.concurrent.ExecutionException {
+		final CompletableFuture<JobSubmissionResult> jobSubmissionResultCompletableFuture = restClusterClient.submitJob(job);
+
+		final JobSubmissionResult jobSubmissionResult = jobSubmissionResultCompletableFuture.get();
+		return jobSubmissionResult.getJobID();
+	}
+
+	private void killApplicationMaster(final String processName) throws IOException, InterruptedException {
+		final Process exec = Runtime.getRuntime().exec("pkill -f " + processName);
```

Comment:
The job could conditionally call `System.exit(...)` based on files in temporary folders. That should be platform independent strategy.

Reply:
But this should happen on the JobManager because we want to restart the application master and this is a bit trickier. I would leave this as a follow up atm.